Allow reordering questions using the keyboard

susnux opened the pull request:
### Part 1: Adding missing aria role
Resolves: #1263 
`div`s require a role when `aria-label` is set (see context in linked issue)

### Part 2: Add buttons
Resolves: #317 
Added buttons for changing the order. Buttons can be focused and triggered using the keyboard.
This also preserves the drag and drop handling, but also allows using buttons or the keyboard.

The buttons are only shown if hovered / focused, but they are always accessible for screen readers.

If a button is clicked / triggered the focus is kept on the button (the vanishing focus ring in the following video happens only because disabled buttons from `@nextcloud/vue` do not have a visual focus ring)


https://user-images.githubusercontent.com/1855448/220498206-3a1befaf-6def-45e1-9be7-980e8ee4b277.mp4

jancborchardt reviewed (changes requested):
Very nice enhancement! :) Makes it very obvious and accessible how to reorder, good stuff.

Only one thing: When reordering via these buttons, whether it be via keyboard or mouse click, it would be nice to have a quick animation of the movement using `var(--animation-quick)` or `var(--animation-slow)` depending on what looks better. :)

susnux commented:
> Very nice enhancement! :) Makes it very obvious and accessible how to reorder, good stuff.

Thank you :)

> Only one thing: When reordering via these buttons, whether it be via keyboard or mouse click, it would be nice to have a quick animation of the movement using `var(--animation-quick)` or `var(--animation-slow)` depending on what looks better. :)

Done. Added the animation, I choose `slow` because `quick` for long questions looks too "jumpy".


https://user-images.githubusercontent.com/1855448/220657783-6cda0693-8d11-4d95-baa1-de646dce0d8f.mp4

jancborchardt reviewed (approved):
Looks really amazing now @susnux, great work! :)

Chartman123 commented:
@susnux I get lots of errors in the browser console like this:

```
[Vue warn]: Invalid prop: type check failed for prop "setReturnFocus". Expected Object, String, Function, Boolean, got HTMLButtonElement 

found in

---> <NcPopover>
       <NcActions>
         <RouterLink>
           <NcListItem>
             <AppNavigationForm> at src/components/AppNavigationForm.vue
               <NcAppNavigation>
                 <NcContent>
                   <Forms> at src/Forms.vue
                     <Root> 2 forms-main.js:134698:21
```

The lower part of the message varies...

And at least for me the focus ring always stays on the ⬇️ button after clicking on it with the mouse. The ⬆️ button doesn't show the ring.

susnux commented:
@Chartman123 the errors should be fixed with nextcloud Vue 7.7.0:
https://github.com/nextcloud/nextcloud-vue/pull/3803

Chartman123 commented:
@susnux: With nc-vue 7.7 the errors are gone. But now I found some new ones: canMoveUp and canMoveDown are missing in the view mode and thus throwing errors.

```
[Vue warn]: Missing required prop: "canMoveUp"

found in

---> <Question> at src/components/Questions/Question.vue
       <QuestionDropdown> at src/components/Questions/QuestionDropdown.vue
         <NcAppContent>
           <Submit> at src/views/Submit.vue
             <NcContent>
               <Forms> at src/Forms.vue
                 <Root>
```

susnux commented:
> With nc-vue 7.7 the errors are gone. But now I found some new ones: canMoveUp and canMoveDown are missing in the view mode and thus throwing errors.

@Chartman123 those issues are resolved. And the focus issue is also resolved.

Chartman123 commented:
> > With nc-vue 7.7 the errors are gone. But now I found some new ones: canMoveUp and canMoveDown are missing in the view mode and thus throwing errors.
> 
> @Chartman123 those issues are resolved. And the focus issue is also resolved.

Nice :) I will check it this evening

Chartman123 commented on src/components/Questions/Question.vue:
```diff
@@ -31,8 +31,29 @@
-			:aria-label="t('forms', 'Drag to reorder the questions')">
+			:aria-label="t('forms', 'Drag to reorder the questions')"
+			role="button">
+			<NcButton ref="buttonU"
```
```suggestion
			<NcButton ref="buttonUp"
```
shouldn't this be buttonUp?

Chartman123 commented on src/components/Questions/Question.vue:
```diff
@@ -218,6 +253,18 @@ export default {
+		},
+		onMoveUp() {
+			this.$emit('move-up')
+			this.$nextTick(() => this.$refs.buttonU.$el.focus())
```
```suggestion
			this.$nextTick(() => this.$refs.buttonUp.$el.focus())
```
Same here

Chartman123 commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -30,10 +30,10 @@
 		@update:text="onTitleChange"
 		@update:description="onDescriptionChange"
-		@update:isRequired="onRequiredChange"
-		@delete="onDelete">
```
Can we remove the onDelete method from Question.vue if you remove it from every question type?

susnux replied:
No. We still need to emit the `delete` event from the `Question`.

I just deleted the `onDelete` from the "parent" questions (like `QuestionDate`), as it is already included within the `v-on="$listeners"`.

jotoeri commented on src/views/Create.vue:
```diff
@@ -85,15 +85,21 @@
-					:max-string-lengths="maxStringLengths"
-					v-bind.sync="form.questions[index]"
-					@delete="deleteQuestion(question)" />
+				<transition-group :move-class="isDragging ? 'nomove' : undefined">
```
Any reason for that nomove? A transition was desired here (see 200ms on Draggable). - Even that the old transition seems to get broken with this PR, too. 😕

susnux replied:
If you drag it, a transition looks quite bad as when you drop the question you would expect it to be just where you dropped it and not on its previous position and moving to the new.

jotoeri replied:
That's true - but while dragging the transition should be maintained as it is on master currently.

susnux replied:
Fixed

jancborchardt commented on src/components/Questions/Question.vue:
```diff
@@ -31,8 +31,29 @@
 			class="question__drag-handle"
 			:class="{'question__drag-handle--shiftup': shiftDragHandle}"
-			:aria-label="t('forms', 'Drag to reorder the questions')">
+			:aria-label="t('forms', 'Drag to reorder the questions')"
+			role="button">
```
cc accessibility experts @JuliaKirschenheuter @Pytal – what would you recommend?

Pytal replied:
Since reordering can be done via keyboard with the move buttons already then there is already feature parity for keyboard-only users

We can still keep the draggable functionality for mouse users but the `role` and `aria-label` should be removed as they are never read by screenreaders anyways with the non-focusable `div`

susnux replied:
:+1: removed

susnux commented:
- Rebased
- Fixed a11y issues
- Fixed transition for both: Keyboard and mouse sorting


https://github.com/nextcloud/forms/assets/1855448/935cb36d-28c6-46c8-b64d-9b64e359fbe0

susnux commented on src/components/Questions/Question.vue:
```diff
@@ -319,19 +373,32 @@ export default {
 		display: flex;
 		inset-inline-start: 0;
+		flex-direction: column;
+		justify-content: space-around;
```
I agree, but then we need a `gap` otherwise they are too close. I will go for `12px` :)
